src/bootstrap/dist.rs

```diff
@@ -2297,6 +2297,48 @@ impl Step for RustDev {
     }
 }
 
+// Tarball intended for internal consumption to ease rustc/std development.
+//
+// Should not be considered stable by end users.
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct RustDevConfig {
+    pub target: TargetSelection,
+}
+
+impl Step for RustDevConfig {
+    type Output = Option<GeneratedTarball>;
+    const DEFAULT: bool = true;
+    const ONLY_HOSTS: bool = true;
+
+    fn should_run(run: ShouldRun<'_>) -> ShouldRun<'_> {
+        run.alias("rust-dev-config")
+    }
+
+    fn make_run(run: RunConfig<'_>) {
+        run.builder.ensure(RustDevConfig { target: run.target });
+    }
+
+    fn run(self, builder: &Builder<'_>) -> Option<GeneratedTarball> {
+        let target = self.target;
+
+        /* run only if llvm-config isn't used */
+        if let Some(config) = builder.config.target_config.get(&target) {
+            if let Some(ref _s) = config.llvm_config {
+                builder.info(&format!("Skipping RustDevConfig ({}): external LLVM", target));
+                return None;
+            }
+        }
+
+        let mut tarball = Tarball::new(builder, "rust-dev-config", &target.triple);
+        tarball.set_overlay(OverlayKind::LLVM);
+
+        let config = t!(serde_json::to_string_pretty(&builder.build.config.llvm));
+        t!(std::fs::write(tarball.image_dir().join("llvm-opts.json"), config));
+
+        Some(tarball.generate())
```

Comment:
this panics when running `x dist rust-dev-config` before the llvm submodule is checked out. could you add `builder.update_submodule` somewhere around https://github.com/rust-lang/rust/blob/0a1b9834b3d71953ff0b22b48237e23d5fa9fd3f/src/bootstrap/tarball.rs#L305-L307 please?

Reply:
Should I really update LLVM submodule in a function that creates a tarball? Shouldn't this be in `RustDevConfig`? It sounds quite LLVM specific to do this in the `Tarball` struct.

Reply:
hmm, i suppose doing it in this Step is fine. i was imaging you'd check overlay_kind in Tarball so each calling Step doesn't have to worry about it but in practice they're probably all doing the right thing.